### PR TITLE
Advanced Building Blocks Enumerables

### PR DESCRIPTION
**Pull request by JohnHernCode**
Project #2 Microverse Ruby Path - Advanced Building Blocks Enumerables

The goal of this project was to recreate enumerables built in methods with plain ruby.

Milestones:
1. Recreate Each
2. Recreate Each with Index
3. Recreate Select
4. Recreate All?
5. Recreate Any?
6. Recreate None?
7. Recreate Count
8. Recreate Map (also include proc and default to proc if proc and block are input)
9. Recreate Inject (also test inject with multiply_els method)

### CODE REVIEW

**Review by umairarshadbutt — COMMENTED** [review]
.

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @JohnHernCode  and @gscarv13 
Such a remarkable project you made :+1: ! Awesome README :100: !
We, however, do have a few changes to carry out before we proceed (see review comment).

##

 For this project, work on the suggested changes  👇.
- [x] Rename the file name from `main.rb` to `enumerables.rb`
- [x] Remove `puts` and `p` from ruby code.
## NOTE: These are the variables I used in the tests below: 👇 
```
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9

describe 'enumerables' do
  let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }
  let(:words) { %w[dog door rod blade] }
  let(:range) { Range.new(5, 50) }
  let(:hash) { { a: 1, b: 2, c: 3, d: 4, e: 5 } }
  let(:numbers) { [1, 2i, 3.14] }
  let!(:array_clone) { array.clone }
```
## Failures: 44
- [x] #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in self
### Test 👉 `expect(range.my_each(&block)).to eq(range.each(&block))`
|Error Name|
|:---:|
|` undefined local variable or method `length' for 5..50:Range`|



- [x] #my_each_with_index when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
### Test 👉 `enum.my_each_with_index(&block)`
     

       undefined local variable or method `length' for 5..50:Range
     Shared Example Group: "#each_with_index" called from ./spec/enumerables_spec.rb:102

- [x] enumerables #my_select returns an array containing all elements of enum for which the given block returns a true value::range
### Test 👉 `expect(range.my_select(&block)).to eq(range.select(&block))`
     

       undefined local variable or method `length' for 5..50:Range


- [x] enumerables #my_select returns an enumerator if no block is given
     Failure/Error: expect(array.my_select).to be_an(Enumerator)
     

       no block given (yield)


- [x] #my_all? returns true if the block never returns false or nil
### Test 👉 `let(:true_block) { proc { |num| num <= HIGHEST_VALUE } }`
 
     NoMethodError:
       undefined method `<=' for false:FalseClass
       Did you mean?  <=>

- [x] #my_all? returns false if the block returns false or nil
### Test 👉 `let(:false_block) { proc { |num| num > HIGHEST_VALUE } }`

     NoMethodError:
       undefined method `>' for false:FalseClass

- [x] #my_all? returns false if the block returns false or nil::range
 ### Test 👉 ` expect(range.my_all?(&false_block)).to eq(range.all?(&false_block))`

     NameError:
       undefined local variable or method `length' for 5..50:Range

- [x] enumerables #my_all? does not mutate the original array
### Test 👉 `array.my_all? { |num| num + 1 }`

      NoMethodError:
        undefined method `+' for false:FalseClass

- [x] #my_all? when no block or argument is given returns true when none of the collection members are false or nil
### Test 👉 `expect(true_array.my_all?).to be true_array.all?`

      LocalJumpError:
        no block given (yield)

- [x] #my_all? when no block or argument is given returns false when one of the collection members are false or nil
### Test 👉 `expect(false_array.my_all?).to be false_array.all?`

      LocalJumpError:
        no block given (yield)

- [x] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
### Test 👉 `expect(array.my_all?(Integer)).to be array.all?(Integer)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
### Test 👉 `expect(numbers.my_all?(Numeric)).to be numbers.all?(Numeric)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)


- [x] #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
### Test 👉 `expect(array.my_all?(Integer)).to be array.all?(Integer)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
### Test 👉 `expect(words.my_all?(/d/)).to be words.all?(/d/)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex
### Test 👉 `expect(words.my_all?(/o/)).to be words.all?(/o/)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
### Test 👉 `expect(array.my_all?(3)).to be array.all?(3)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
### Test 👉 `expect(array.my_all?(3)).to be array.all?(3)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_any? returns false if the block ever returns a value false or nil::range
### Test 👉 `expect(range.my_any?(&false_block)).to eq(range.any?(&false_block))`

      NameError:
        undefined local variable or method `length' for 5..50:Range

- [x] #my_any? when no block or argument is given returns false if at least one of the collection is not true
### Test 👉 `expect(false_array.my_any?).to be false_array.any?`

        expected false
             got true

- [x] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
### Test 👉 `expect(array.my_any?(Numeric)).to be array.any?(Numeric)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
### Test 👉 `expect(words.my_any?(Integer)).to be words.any?(Integer)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
### Test 👉 `expect(words.my_any?(/d/)).to be words.any?(/d/)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
### Test 👉 `expect(words.my_any?(/z/)).to be words.any?(/z/)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
### Test 👉 `expect(words.my_any?('cat')).to be words.any?('cat')`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
### Test 👉 `expect(words.my_any?('cat')).to be words.any?('cat')`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_none? returns false if the block ever returns true for all elements::range
### Test 👉 `expect(range.my_none?(&false_block)).to eq(range.none?(&false_block))`

      NameError:
        undefined local variable or method `length' for 5..50:Range

- [x] #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
### Test 👉 `expect(array.my_none?(String)).to be array.my_none?(String)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
### Test 👉 `expect(array.my_none?(Numeric)).to be array.my_none?(Numeric)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_none? when a class is passed as an argument returns false if any of the collection is a member of such class
### Test 👉 `expect(array.my_none?(String)).to be array.my_none?(String)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
### Test 👉 `expect(words.my_none?(/z/)).to be words.none?(/z/)`
      |Error Name|

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
### Test 👉 `expect(words.my_none?(/d/)).to be words.none?(/d/)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
### Test 👉 `expect(words.my_none?(5)).to be words.none?(5)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
### Test 👉 `expect(words.my_none?(5)).to be words.none?(5)`
      |Error Name|

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_count returns the number of items in enum through enumeration::range
### Test 👉 `expect(range.my_count).to eq range.count`

      NameError:
        undefined local variable or method `length' for 5..50:Range

- [x] #my_count counts the number of items in enum that are equal to item if an argument is given
### Test 👉 `expect(array.my_count(LOWEST_VALUE)).to eq array.count(LOWEST_VALUE)`
     
        expected: 11
             got: 100
      
        (compared using ==)

- [x] #my_inject searches for the longest word in an array of strings
### Test 👉 `let(:search) { proc { |memo, word| memo.length > word.length ? memo : word } }`

      NoMethodError:
        undefined method `length' for 1:Integer

- [ ] #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
### Test 👉 `expect(array.my_inject(&operation)).to eq array.inject(&operation)`

        expected: 425
             got: 426
      

- [x] #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
### Test 👉 `actual = range.my_inject(4) { |prod, n| prod * n }`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
### Test 👉 `expect(array.my_inject(:+)).to eq array.inject(:+)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
### Test 👉 `expect(range.my_inject(:*)).to eq range.inject(:*)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)

- [x] #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
### Test 👉 `expect(array.my_inject(20, :*)).to eq array.inject(20, :*)`

      ArgumentError:
        wrong number of arguments (given 2, expected 0)

- [x] #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
### Test 👉 `expect(range.my_inject(2, :*)).to eq range.inject(2, :*)`

      ArgumentError:
        wrong number of arguments (given 2, expected 0)
        
##
Tip: Please explore [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each), [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index), [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select), [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F), [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F), [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F), [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count), [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map), [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject)
##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)
##

**Comment by gscarv13** [comment]
# https://emojipedia.org/lady-beetle/ 🐞

**Comment by JohnHernCode** [comment]
### 🐞☠️  Variables not provided ☠️  🐞
The following case was not able to be tested as the variables were not provided. 

Outside of this, all other tests were complete. 

###  👉 Test Case Without Variables Below

> - [ ] #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
> ### Test 👉 `expect(array.my_inject(&operation)).to eq array.inject(&operation)`

>        expected: 425
>             got: 426

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: #my_all? when no block or argument is given returns true when none of the collection members are false or nil
#my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: #my_any? when no block or argument is given returns false if at least one of the collection is not true
#my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

File: enumerables.rb
Comment (javitocor) [inline enumerables.rb]: #my_inject raises a "LocalJumpError" when no block or argument is given
#my_inject searches for the longest word in an array of strings
#my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
#my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED ♻️ 

Hello team! 👋 
A great effort implementing the suggested changes 👍 
However, we still need to make some changes in order to get this project approved 💪 
Kindly check my comments below 👇  and request a new code review once done ⏩ 
Good luck! 😄

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've done really well so far.
However, some changes are required before this project can be approved. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**

- [ ] Using the same dataset and test cases used by the previous reviewers, we have just 2 failed test cases, coming from `#my_any?` and `#my_inject` methods. Let's check them out!
- When no block or argument is given, `#my_any? should return true if at least one of the collections is not false or nil. At the moment, it returned false.
- #my_inject should raise a `LocalJumpError` when no block or argument is given. Right now, `<NoMethodError: undefined method `call' for nil:NilClass>` was raised.

**That's it!**
* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED ♻️ 

Hello team! 👋 
Almost there! Just one more issue to fix and ready to go! 🚀 
Kindly check my comments below 👇  and resubmit your project once the changes  are done ⏩ 
Good luck! 😺 
## Mandatory changes
- #my_any? when no block or argument is given returns false if at least one of the collection is not true

**Review by mcrd25 — APPROVED** [review]
# Changes Approved :heavy_check_mark: 

Hi Team! :wave: 

Thanks so much for making the changes! all seems good now! :smile: 

***If you have any questions, now or in the future, do not hesitate to contact me on social media, email or on slack and I will be glad to chat with you.***

If you liked my review, consider upskilling me on LinkedIn. And please do connect. [My LinkedIn](https://www.linkedin.com/in/mayadouglas/)


*Happy coding 😄,*
**[--Maya Douglas](mayacdouglas.com)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

**Pull request merged**
